Asynchronous halo exchanges #214

rupertford opened the pull request:
Ready for first review

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -2203,7 +2208,7 @@ def field(self):
     def dag_name(self):
         ''' Return the name to use in a dag for this node'''
-        name = ("haloexchange({0})_".format(self._field.name) +
+        name = ("{0}({1})_".format(self._dag_name, self._field.name) +
```
Might as well be consistent and use format for all three:
```suggestion
        name = ("{0}({1})_{2}".format(self._dag_name, self._field.name, self.position)
```

rupertford replied:
Agreed.

arporter commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -6474,3 +6475,546 @@ def test_no_acc():
+def test_async_hex_wrong_node():
+    '''Test that we raise the expected exception if an asynchronous halo
+    exchange transformation is applied to a node that is not a halo
+    exchange
```
Full stop please. (I blame @TeranIvy :-))

rupertford replied:
What goes around comes around :-)

arporter commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -6474,3 +6475,546 @@ def test_no_acc():
+def test_async_hex_move_error_2():
+    '''Test that an asynchronous halo exchange start can not be moved
+    before a kernel that modifies the field and its end cannot be
+    moved after a kernel that reads the field
```
Full stop please.

rupertford replied:
Right, I've added missing full stops to all test descriptions. There were quite a few needed doing.

arporter commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -6474,3 +6475,546 @@ def test_no_acc():
+
+
+def test_rc_remove_async_halo_exchange(monkeypatch, tmpdir, f90, f90flags):
+    '''Test that a halo exchange is removed if redundant computation means
```
"an asynchronous halo exchange is ..."

rupertford replied:
Done.

arporter commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -6474,3 +6475,546 @@ def test_no_acc():
+    assert "CALL m1_proxy%halo_exchange(depth=1)" in result
+    #
+
+    invoke = psy.invokes.invoke_list[0]
```
You've already got a reference to `schedule` above.

rupertford replied:
Removed.

arporter commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -6474,3 +6475,546 @@ def test_no_acc():
+    assert "CALL f2_proxy%halo_exchange_finish(depth=1)" in result
+    assert "IF (m1_proxy%is_dirty(depth=1)) THEN" in result
+    assert "CALL m1_proxy%halo_exchange(depth=1)" in result
+    #
```
Please delete this line.

rupertford replied:
Removed.

arporter commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -6474,3 +6475,546 @@ def test_no_acc():
+        "      END IF \n"
+        "      !\n"
+        "      IF (m2_proxy%is_dirty(depth=2)) THEN\n"
+        "        CALL m2_proxy%halo_exchange_finish(depth=2)\n"
```
I guess this is the LFRic API rather than PSyclone but it seems strange that we have to pass in `depth` to `halo_exchange_finish`.

rupertford replied:
Yes it is their API. I hadn't really thought about it, but could knowing the size of data being received e.g. to allow pre-allocation of appropriate buffer space could potentially improve performance?

arporter commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -6474,3 +6475,546 @@ def test_no_acc():
+def test_move_vector_halo_exchange():
+    '''Test that halo exchanges for different vectors for the same field
+    are independent of each other, i.e. they do not depend on
+    eachother
```
"each other."

rupertford replied:
All eachothers changed to each other.

arporter commented on src/psyclone/tests/expression_test.py:
```diff
@@ -22,13 +22,13 @@ def my_test(name, parser, test_string, names=None):
+        'NamedArg': psyclone.expression.NamedArg,
+        'Slicing': psyclone.expression.Slicing,
+    }
     # pylint: disable=exec-used
```
Tidying for pylint? Thanks.

rupertford replied:
Yes.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -2206,6 +2206,96 @@ def apply(self, nodes):
+        '''
+        from psyclone.psyGen import HaloExchange
+
+        if not isinstance(node, HaloExchange):
```
What happens if `node` is a `DynHaloExchangeStart` or `End`? I think you need to test explicitly for that.

rupertford commented on doc/developers.rst:
```diff
@@ -804,6 +804,61 @@ determined by the halo exchange itself at code generation time. The
+Aynschronous Halo Exchanges
++++++++++++++++++++++++++++
+
+The Dynamo0p3AsynchronousHaloExchange transformation allows the
```
I hear you but all transformations are written this way at the moment so changing this one would make it inconsistent with the others and the associated documentation. There is also the related question of how to ensure that we only show appropriate transformations for a particular API when they are listed. So I think what you are requesting is part of a bigger restructuring.

rupertford commented on doc/developers.rst:
```diff
@@ -804,6 +804,61 @@ determined by the halo exchange itself at code generation time. The
+Both halo exchange start and halo exchange end inherit from halo
+exchange. However, the halo exchange start and end are really two
+parts of the same thing and need to have consistent properties
+including after transformations are performed. This is achieved by
```
Done.

rupertford commented on doc/developers.rst:
```diff
@@ -804,6 +804,61 @@ determined by the halo exchange itself at code generation time. The
+and this could probably be removed at the expense of returning
+appropriate names for the dag, colourmap, declaration etc.
+
+.. note:: Halo exchange vectors are currently broken when an existing
```
Modified, thanks.

rupertford commented on doc/developers.rst:
```diff
@@ -804,6 +804,61 @@ determined by the halo exchange itself at code generation time. The
+   of them are removed. This is also the case for asynchronous halo
+   exchanges. See issue #219.
+
+.. note:: The dependence analysis for vector is currently over
```
Better, thanks.

rupertford commented on examples/dynamo/eg11/apply_variable_hx_kernel_mod.F90:
```diff
@@ -0,0 +1,346 @@
+!-----------------------------------------------------------------------------
+! (C) Crown copyright 2017 Met Office. All rights reserved.
+! The file LICENCE, distributed with this code, contains details of the terms
```
OK, I've removed the file as it is not needed for the example.

rupertford commented on examples/dynamo/eg11/async_script.py:
```diff
@@ -0,0 +1,39 @@
+def trans(psy):
```
Oops. Done.

rupertford commented on examples/dynamo/eg11/async_script.py:
```diff
@@ -0,0 +1,39 @@
+def trans(psy):
+    '''sample transformation script to demonstrate the use of asynchronous
```
Yes sir.

rupertford commented on examples/dynamo/eg11/async_script.py:
```diff
@@ -0,0 +1,39 @@
+def trans(psy):
+    '''sample transformation script to demonstrate the use of asynchronous
+    halo exchanges with overlapping compute and communication for the
+    most costly halo exchanges in the (current version of the) model
```
Quite right.

rupertford commented on examples/dynamo/eg11/async_script.py:
```diff
@@ -0,0 +1,39 @@
+
+    # This transformation removes the halo exchange associated with
+    # the grad_p field. This transformation is not be needed if
+    # annexed_dofs is set to True in the config file.
```
Changed typo.

I'm not sure about checking annexed dofs. The code is still correct, we just change the bounds to halo(1) and that removes a halo exchange when annexed_dofs is False and does nothing when annexed dofs is True (apart from performing redundant computation into the level 1 halo rather than up to and including annexed dofs).

arporter replied:
OK. As long as it works in both cases I'm happy.

rupertford commented on examples/dynamo/eg11/async_script.py:
```diff
@@ -0,0 +1,39 @@
+    # (for fields p, hb_inv and u_normalisation) into asynchronous
+    # (halo_exchange_start and halo_exchange_end) ones.
+    ahex_trans = Dynamo0p3AsyncHaloExchangeTrans()
+    for index in [3, 2, 1]:
```
OK, if you insist, although I don't see n advantage.

rupertford commented on examples/dynamo/eg11/async_script.py:
```diff
@@ -0,0 +1,39 @@
+    # before the setval_c loop offering the potential for overlap
+    # between communication and computation.
+    mtrans = MoveTrans()
+    for index in [5, 4, 3]:
```
Same thought.

rupertford commented on examples/dynamo/eg11/helmholtz_solver_alg_mod.x90:
```diff
@@ -0,0 +1,40 @@
+
```
Done.

rupertford commented on examples/dynamo/eg11/scaled_matrix_vector_kernel_mod.F90:
```diff
@@ -0,0 +1,201 @@
+!-----------------------------------------------------------------------------
+! (C) Crown copyright 2017 Met Office. All rights reserved.
+! The file LICENCE, distributed with this code, contains details of the terms
```
Done (thanks to copying eg10!)

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3248,7 +3247,27 @@ class DynHaloExchange(HaloExchange):
+
+    :param field: the field that this halo exchange will act on
+    :type field: :py:class:`psyclone.dynamo0p3.DynKernelArgument`
+    :param check_dirty: optional argument default True indicating
```
Done.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3248,7 +3247,27 @@ class DynHaloExchange(HaloExchange):
+    identify which index of a vector field this halo exchange is
+    responsible for
+    :type vector_index: int
+    :param parent: optional parent (default None) of this object
```
OK. First use of the PSyIRe in our documentation (by me at least).

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3534,15 +3553,24 @@ def view(self, indent=0):
         runtime_check = not known
+        field_id = self._field.name
+        if self.vector_index:
+            field_id += "({0})".format(self.vector_index)
```
Thanks. The PSyIRe output from view() makes much more sense with this in.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3560,10 +3588,162 @@ def gen_code(self, parent):
+                 vector_index=None, parent=None):
+        DynHaloExchange.__init__(self, field, check_dirty=check_dirty,
+                                 vector_index=vector_index, parent=parent)
+        if field:
```
No I don't think it makes sense to be None. The base class allows it to be None which is why I did the same here but there is no way of setting it once a halo exchange has been created (other than accessing the internal class variables directly) so I don't know why I did that. I've therefore removed the 'if' test.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3560,10 +3588,162 @@ def gen_code(self, parent):
+        only read and the dependence analyis beneath this call
+        requires the field to be modified.
+
+        :return: Return the type of stencil required for this pair of
```
OK.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3560,10 +3588,162 @@ def gen_code(self, parent):
+    def _compute_stencil_type(self):
+        '''Call the required method in the corresponding halo exchange end
+        object. This is done as the field in halo exchange start is
+        only read and the dependence analyis beneath this call
```
Well spotted.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3560,10 +3588,162 @@ def gen_code(self, parent):
+        object or raises an exception if one is not found.
+
+        :return: The corresponding halo exhange end object
+        :rtype::py:class:`psyclone.dynamo0p3.DynHaloExchangeEnd`
```
Modified, thanks.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3560,10 +3588,162 @@ def gen_code(self, parent):
+
+        '''
+        remote_node = self._get_hex_end()
+        return remote_node._compute_stencil_type()
```
OK, done.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3560,10 +3588,162 @@ def gen_code(self, parent):
+    def _compute_halo_depth(self):
+        '''Call the required method in the corresponding halo exchange end
+        object. This is done as the field in halo exchange start is
+        only read and the dependence analyis beneath this call
```
Done.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3560,10 +3588,162 @@ def gen_code(self, parent):
+
+        '''
+        remote_node = self._get_hex_end()
+        return remote_node._compute_halo_depth()
```
Done.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3560,10 +3588,162 @@ def gen_code(self, parent):
+    def required(self):
+        '''Call the required method in the corresponding halo exchange end
+        object. This is done as the field in halo exchange start is
+        only read and the dependence analyis beneath this call
```
Done.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3560,10 +3588,162 @@ def gen_code(self, parent):
+
+        '''
+        remote_node = self._get_hex_end()
+        return remote_node.required()
```
Done.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3560,10 +3588,162 @@ def gen_code(self, parent):
+
+        :return: The corresponding halo exhange end object
+        :rtype::py:class:`psyclone.dynamo0p3.DynHaloExchangeEnd`
+        :raises GenerationError: If a matching HaloExchangeEnd is not
```
Picky :-)

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3560,10 +3588,162 @@ def gen_code(self, parent):
+        for node in self.following():
+            if self.sameParent(node):
+                if isinstance(node, DynHaloExchange):
+                    # ************** What about matching field vectors?????
```
I've created a TODO. It's not a biggy as we get correct results for field vectors with the existing logic. (as all haloexchangeend nodes associated with a field vector should return the same results).

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3560,10 +3588,162 @@ def gen_code(self, parent):
+
+        '''
+        for node in self.following():
+            if self.sameParent(node):
```
Done.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3560,10 +3588,162 @@ def gen_code(self, parent):
+        not a HaloExchangeEnd
+
+        '''
+        for node in self.following():
```
OK, added some words.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3560,10 +3588,162 @@ def gen_code(self, parent):
+                    if node.field.name == self.field.name:
+                        if isinstance(node, DynHaloExchangeEnd):
+                            return node
+                        else:
```
It's for the case when it is not a haloexchangeend e.g. it might match with a haloexchange, or another haloexchange start which would be an error.

arporter replied:
I meant that you don't need the `else` itself:

    if isinstance(node, blah):
        return node
    raise GenerationError("blah")

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3560,10 +3588,162 @@ def gen_code(self, parent):
+            if self.sameParent(node):
+                if isinstance(node, DynHaloExchange):
+                    # ************** What about matching field vectors?????
+                    if node.field.name == self.field.name:
```
I think I've addressed this.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3560,10 +3588,162 @@ def gen_code(self, parent):
+
+    :param field: the field that this halo exchange will act on
+    :type field: :py:class:`psyclone.dynamo0p3.DynKernelArgument`
+    :param check_dirty: optional argument default True indicating
```
Done.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3560,10 +3588,162 @@ def gen_code(self, parent):
+    identify which index of a vector field this halo exchange is
+    responsible for
+    :type vector_index: int
+    :param parent: optional parent (default None) of this object
```
Done.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3560,10 +3588,162 @@ def gen_code(self, parent):
+                 vector_index=None, parent=None):
+        DynHaloExchange.__init__(self, field, check_dirty=check_dirty,
+                                 vector_index=vector_index, parent=parent)
+        if field:
```
I've removed this.

rupertford commented on src/psyclone/psyGen.py:
```diff
@@ -103,12 +103,14 @@ def colored(text, _):
 # coloured.) See https://pypi.python.org/pypi/termcolor for details.
-SCHEDULE_COLOUR_MAP = {"Schedule": "yellow",
-                       "Loop": "white",
+SCHEDULE_COLOUR_MAP = {"Schedule": "white",
```
Thanks. It does look better.

rupertford commented:
I've hopefully addressed the reviewers comments and have also brought the PR up-to-date with the latest version of master. Back to @arporter for 2nd review.

arporter commented on doc/dynamo0p3.rst:
```diff
@@ -2031,10 +2032,12 @@ the spaces are the same. This option should therefore be used with
+"Dynamo0p3" API. This is because this API is currently the only one
+that supports distributed memory.  An example of redundant computation
+can be found in ``examples/dynamo/eg8`` and an example of asynchronous
+halo exchanges can be found in ``examples/dynamo/eg11``.
```
:-)

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3841,29 +3859,28 @@ class DynHaloExchangeEnd(DynHaloExchange):
     :type field: :py:class:`psyclone.dynamo0p3.DynKernelArgument`
-    :param check_dirty: optional argument default True indicating
-    whether this halo exchange should be subject to a run-time check
+    :param check_dirty: optional argument default True indicating \
```
s/default True/(default True)/

rupertford replied:
Done.

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3841,29 +3859,28 @@ class DynHaloExchangeEnd(DynHaloExchange):
-            self._field.access = "gh_readwrite"
+        # Update fields values appropriately. The associated field is
+        # written to. However, a readwrite field access needs to be
+        # specified as this required for the halo exchange logic to
```
s/as this/as this is/

rupertford replied:
Fixed.

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3841,29 +3859,28 @@ class DynHaloExchangeEnd(DynHaloExchange):
-            # probably officially a write but readwrite is needed for
-            # the halo exchange logic to work correctly in any case.
-            self._field.access = "gh_readwrite"
+        # Update fields values appropriately. The associated field is
```
"Update field properties..."?

rupertford replied:
Yes, much better.

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3687,10 +3716,178 @@ def gen_code(self, parent):
+                    if isinstance(node, DynHaloExchangeEnd):
+                        return node
+                    else:
+                        raise GenerationError(
```
This line is not covered (according to CodeCov) and nor is the `raise GenerationError` at line 3849 below.

rupertford replied:
I think my excuse for this was that I was waiting for the vector dependence logic to be fixed first. This has now been fixed and is on master and I've added the missing tests.

rupertford commented on doc/developers.rst:
```diff
@@ -882,19 +882,20 @@ halo exhange start really benefits from inheriting from halo exchange
-   are independent of each other. This is also the case for
-   asynchronous halo exchanges. See issue #220.
+.. note:: There is currently a problem with halo-exchanges for field
+   vectors: when an existing set of halo exchanges associated with a
```
Fixed.

rupertford commented on doc/developers.rst:
```diff
@@ -882,19 +882,20 @@ halo exhange start really benefits from inheriting from halo exchange
+.. note:: The dependence analysis for halo exchanges for field vectors
+   is currently over zealous. It does not allow halo exchanges for
+   independent vector components to be moved past one another. For
+   example, a halo exchange for vector 2, if placed after a halo
```
Updated.

rupertford commented on examples/dynamo/eg11/async_script.py:
```diff
@@ -13,8 +57,9 @@ def trans(psy):
     # This transformation removes the halo exchange associated with
-    # the grad_p field. This transformation is not be needed if
-    # annexed_dofs is set to True in the config file.
+    # the grad_p field. This transformation is unecessary if
```
I found one in developers.rt as well :-)

rupertford commented on examples/dynamo/eg11/async_script.py:
```diff
@@ -13,8 +57,9 @@ def trans(psy):
-    # annexed_dofs is set to True in the config file.
+    # the grad_p field. This transformation is unecessary if
+    # annexed_dofs is set to True in the config file (although the
+    # transformation still works).
```
If you're happy then I'm happy :-)

rupertford commented on examples/dynamo/eg11/async_script.py:
```diff
@@ -23,17 +68,16 @@ def trans(psy):
-        schedule, _ = mtrans.apply(schedule.children[index],
-                                   schedule.children[0])
+    for kern in schedule.children[5:0:-2]:
+        schedule, _ = mtrans.apply(kern, schedule.children[0])
```
OK, I admit it is nicer, although Python not including the last index is annoying imo.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3537,10 +3538,10 @@ def required(self):
-        argument specifies whether we definitely know that we need the
+        :return: Returns (x, y) where x specifies whether this halo \
+        exchange is (or might be) required - True, or is not required \
+        - False. If the first argument is True then the second \
```
I don't think that's picky, it's a good spot.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3723,37 +3724,40 @@ def gen_code(self, parent):
     :type field: :py:class:`psyclone.dynamo0p3.DynKernelArgument`
-    :param check_dirty: optional argument default True indicating
-    whether this halo exchange should be subject to a run-time check
+    :param check_dirty: optional argument default True indicating \
```
Yes, better, thanks.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3723,37 +3724,40 @@ def gen_code(self, parent):
-    :param vector_index: optional vector index (default None) to
-    identify which index of a vector field this halo exchange is
+    :param vector_index: optional vector index (default None) to \
+    identify which index of a vector field this halo exchange is \
```
You like your vector components :-)

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3763,71 +3767,85 @@ def __init__(self, field, check_dirty=True,
-        argument specifies whether we definitely know that we need the
+        :return: Returns (x, y) where x specifies whether this halo \
+        exchange is (or might be) required - True, or is not required \
+        - False. If the first argument is True then the second \
```
Done.

rupertford commented:
I've hopefully addressed all of the reviewers comments. I've also merged in the halo vector fix that was pushed to master. That has allowed me to update associated tests, code and documentation. Ready for 3rd review from @arporter.

arporter commented:
Documentation builds and is good.
CodeCov reports 100% for the diff.
All previous requested changes have been made.
Please bring branch up-to-date with master and ensure that all the examples work (I think they will).
There is one test that fails to compile because of problems in `test_files/dynamo0p3/testkern_vector_2_mod.f90`
The BigPicture already has a green tick next to "asynchronous halo exchanges"!?
An example has been added showing the new functionality.
Once the examples and tests (with compilation) all work, I'll merge this to master.

arporter commented:
Tests all pass/xfail with compilation enabled. Examples all run OK.
Codecov still reports 100% coverage of the diff.
Documentation builds and is good.
